Marketplace: Advertise the discount on annual plans

lupus2k opened the pull request:
#### Changes proposed in this Pull Request
- Create a component to get the discount price on annual plans
- Use the component in monthly/yearly selector when in plugin view
- Add styles to match [this requirements](https://github.com/Automattic/wp-calypso/issues/59397#issuecomment-1039792675)
- Had to modify the [`billing-interval-switcher`](https://github.com/Automattic/wp-calypso/tree/trunk/client/my-sites/marketplace/components/billing-interval-switcher) component in order to receive the plugin to get the discount

#### Testing instructions
- Visit /plugins page and select a paid plugin.
- You should see the discount price in the annual selector

#### Screenshots Desktop
Before
<img width="1064" alt="image" src="https://user-images.githubusercontent.com/402286/155017939-61cb0b59-54e2-4a8c-b483-024855864f9f.png">

After
<img width="1056" alt="image" src="https://user-images.githubusercontent.com/402286/155018112-b766898f-8292-4d42-ad79-9f91ade80dd8.png">

#### Screenshots Mobile
Before
<img width="264" alt="image" src="https://user-images.githubusercontent.com/402286/155018010-5a6a5a81-0585-4ca1-b2bd-99876f1357a9.png">

After
<img width="266" alt="image" src="https://user-images.githubusercontent.com/402286/155018232-cd2287c3-9071-4d56-97b3-35aaa6eeacca.png">

Related to #59397

lupus2k committed:
Marketplace: Advertise the discount on annual plans

obenland commented on client/my-sites/marketplace/components/billing-interval-switcher/index.tsx:
```diff
@@ -48,6 +65,21 @@ const BillingIntervalSwitcher: React.FunctionComponent< Props > = ( {
 						onClick={ () => onChange( IntervalLength.ANNUALLY ) }
 					>
 						{ annualLabel }
+						{ plugin ? (
```
Can we check for `plugin?.saving` her or something so we don't have to do multiple checks?

Not sure if the `&&` notation would work here instead of the ternary?
`{ plugin && ( <PluginAnnualSaving /> ) }`

Same below.

lupus2k replied:
I fixed it in [this commit](https://github.com/Automattic/wp-calypso/pull/61341/commits/35806f3d082ede6f7e5577351afd86440f1334d6)

obenland commented on client/my-sites/marketplace/components/billing-interval-switcher/index.tsx:
```diff
@@ -57,7 +89,25 @@ const BillingIntervalSwitcher: React.FunctionComponent< Props > = ( {
+							annualSaving.saving ? (
+								<>
+									<PluginAnnualSavingLabelDesktop>
+										(
```
These wrapping parens can probably be omitted

lupus2k replied:
Removed in [this commit](https://github.com/Automattic/wp-calypso/pull/61341/commits/35806f3d082ede6f7e5577351afd86440f1334d6).

obenland commented on client/my-sites/marketplace/components/billing-interval-switcher/index.tsx:
```diff
@@ -57,7 +89,25 @@ const BillingIntervalSwitcher: React.FunctionComponent< Props > = ( {
+					<PluginAnnualSaving plugin={ plugin }>
+						{ ( annualSaving: { saving: any } ) =>
+							annualSaving.saving ? (
+								<>
```
Is this wrapper needed within `<PluginAnnualSaving />`?

lupus2k replied:
Not needed, It was removed in [this commit](https://github.com/Automattic/wp-calypso/pull/61341/commits/35806f3d082ede6f7e5577351afd86440f1334d6)

obenland commented on client/my-sites/plugins/plugin-saving/index.jsx:
```diff
@@ -0,0 +1,44 @@
+	} );
+};
+
+export const PluginAnnualSavingPercent = ( { plugin, children } ) => {
```
Is this component used anywhere?

lupus2k replied:
Not yet, I prepared in case we need to use it in the future (all other saving texts, like in plans, use percent)

allilevine replied:
One of [the values we follow for Calypso development](https://wpcalypso.wordpress.com/devdocs/docs/guide/0-values.md) I should have mentioned is iterating quickly and not writing code we don't need yet:

> 1. We optimize for iterating quickly
We aren’t trying to build software around a future we can’t predict. We don’t write code that we may not need. We know that no abstractions are better than bad abstractions.

That said, @SaxonF seems enthusiastic about having a percent option in here. 🙂 Is there one you prefer over the other Saxon?

lupus2k replied:
I removed the price saving percent component

allilevine commented on client/my-sites/marketplace/components/billing-interval-switcher/index.tsx:
```diff
@@ -57,7 +89,25 @@ const BillingIntervalSwitcher: React.FunctionComponent< Props > = ( {
+								<>
+									<PluginAnnualSavingLabelDesktop>
+										(
+										{ sprintf( 'Save %(save)s', {
```
It looks like in [the design](https://github.com/Automattic/wp-calypso/issues/59397#issuecomment-1039792675) there's a little space before `(Save...` on desktop and `-` on mobile, can we add that in?

allilevine replied:
<img width="287" alt="Screen Shot 2022-02-21 at 4 50 20 PM" src="https://user-images.githubusercontent.com/1689238/155030344-37e198a2-c874-4ac5-a2dd-23fca11343dd.png">

lupus2k replied:
Fixed using CSS

allilevine commented:
@lupus2k Thanks for tackling this, it's close! 🎉 I noticed that on certain phone sizes (e.g. iPhone 12/13 mini) the dropdown is almost overlapping the Back button:
<img width="376" alt="Screen Shot 2022-02-21 at 4 53 49 PM" src="https://user-images.githubusercontent.com/1689238/155030719-9f9012f3-4b9a-4ac2-ba1e-abbd3deadf05.png">

Does it look like there's a way to reduce the dropdown width?

lsl commented on packages/plans-grid/src/plans-interval-toggle/index.tsx:
```diff
@@ -88,7 +89,10 @@ const PlansIntervalToggle: React.FunctionComponent< PlansIntervalToggleProps > =
-					<span className="plans-interval-toggle__label">{ annuallyLabel }</span>
+					<span className="plans-interval-toggle__label">
+						{ annuallyLabel } { children }
+					</span>
```
Changing from plain text to an element here is messing with the width of the toggle on mobile.

Before 
![Screenshot 2022-02-22 at 20-39-39 WooCommerce Subscriptions Plugin ‹ Site Title — WordPress com](https://user-images.githubusercontent.com/811776/155105181-f309ffad-41a0-44e9-b40b-25ca46ad5fc6.png)

After
![Screenshot 2022-02-22 at 20-39-14 WooCommerce Subscriptions Plugin ‹ Site Title — WordPress com](https://user-images.githubusercontent.com/811776/155105185-48db8aab-0950-4530-b027-33db4a9b9d18.png)

Looks like it's due to [this hack](https://github.com/Automattic/wp-calypso/blob/485614c6c170d636c2d3f3ed0eb0ea3f98f056dd/client/components/select-dropdown/style.scss#L210) on the [data-bold-text attribute](https://github.com/Automattic/wp-calypso/blob/6c73a904e7d782bd7c6e6cf64617a60d3737f7c6/client/components/select-dropdown/item.jsx#L45) rendering the new label as `Annual price, [object Object]`

![Screenshot_2022-02-22_20-20-20](https://user-images.githubusercontent.com/811776/155105188-4bbc0247-6b82-4c58-a0be-c89e430146cc.png)

To fix, I'd try to override it using local css in the [`plans-interval-toggle` stylesheet](https://github.com/Automattic/wp-calypso/blob/e3465afee05c1e9ea083d585d2c91d940c0174c9/packages/plans-grid/src/plans-interval-toggle/style.scss). I don't think you can easily set a value on the underlying prop.

lupus2k replied:
I fixed this issue using CSS. The `data-bold-text attribute` should be fixed in the future.

lsl commented on client/my-sites/marketplace/components/billing-interval-switcher/index.tsx:
```diff
@@ -57,7 +89,25 @@ const BillingIntervalSwitcher: React.FunctionComponent< Props > = ( {
+							annualSaving.saving ? (
+								<>
+									<PluginAnnualSavingLabelDesktop>
+										(
+										{ sprintf( 'Save %(save)s', {
+											save: annualSaving.saving,
+										} ) }
+										)
```
You can inject variables directly with the `translate` method:

```suggestion
										&nbsp;
										{ translate( 'Save %(save)s', {
											comment: 'Sale price label, ex: Save $51',
											args: { save: annualSaving.saving },
										} ) }
```

And just as a heads up: Usually you will actually want to use `sprintf` + `__` from `useI18n` like in [this example](https://github.com/Automattic/wp-calypso/blob/e862169d398950116a43d2f3528164008e6af091/client/my-sites/woocommerce/landing-page.tsx#L61-L67). We're slowly migrating away from the `translate` method to core wp functions but sticking with the legacy usage that's already in use here is totally fine in this sort of situation.

lsl commented on client/my-sites/plugins/plugin-saving/index.jsx:
```diff
@@ -0,0 +1,44 @@
+			: null;
+	};
+
+	return children( {
+		isFetching,
+		saving: getAnnualPriceSavingText(),
+	} );
```
:open_mouth: Nice inversion of control! Have you heard of [render props](https://reactjs.org/docs/render-props.html)? (TIL!)

lsl reviewed (commented):
Looking good so far, nice work!

lupus2k committed:
Update client/my-sites/marketplace/components/billing-interval-switcher/index.tsx

Add translation to Saving price text

Co-authored-by: Louis Laugesen <louis.laugesen@gmail.com>

lupus2k committed:
Marketplace: Advertise the discount on annual plans
- Remove PluginAnnualSavingPercent component
- Remove unused wrapping parens.
- Add class to fix yearly/monthly selector on mobile.

sirbrillig commented on client/my-sites/marketplace/components/billing-interval-switcher/index.tsx:
```diff
@@ -1,9 +1,14 @@
 import { PlansIntervalToggle } from '@automattic/plans-grid/src';
 import styled from '@emotion/styled';
 import { useTranslate } from 'i18n-calypso';
+import React from 'react';
```
Really minor: If this is being imported for the `React. FunctionComponent` type, it could instead be changed to a type-only import (that will be stripped out by the compiler) by changing it to:

```
import type { FunctionComponent } from 'react';
```

And then changing all uses of `React.FunctionComponent` to just `FunctionComponent`.

lupus2k replied:
Fixed [here](https://github.com/Automattic/wp-calypso/pull/61341/commits/cbe79a5a646d9b0280938f2b7d1a64405769d753)

lupus2k committed:
Marketplace: Advertise the discount on annual plans
- Update type-only import for React.FunctionComponent

sirbrillig commented on client/my-sites/marketplace/components/billing-interval-switcher/index.tsx:
```diff
@@ -1,9 +1,14 @@
 import { PlansIntervalToggle } from '@automattic/plans-grid/src';
 import styled from '@emotion/styled';
 import { useTranslate } from 'i18n-calypso';
+import { FunctionComponent } from 'react';
```
```suggestion
import type { FunctionComponent } from 'react';
```

sirbrillig commented on client/my-sites/marketplace/components/billing-interval-switcher/index.tsx:
```diff
@@ -16,16 +21,28 @@ const PlansIntervalToggleLabel = styled.span`
 	onChange: ( selectedValue: 'MONTHLY' | 'ANNUALLY' ) => void;
 	billingPeriod: IntervalLength;
 	compact: boolean;
+	plugin?: object;
```
Using `object` as a type is not ideal, since it doesn't tell us anything about the data. If you don't know what the data is, you can use `unknown`. However, in this case I think we do know what it is, since I see it being used below like `plugin?.variations?.yearly?.product_slug`.

```suggestion
	plugin?: { variations?: { yearly?: { product_slug: string }; monthly?: { product_slug: string } } };
```

sirbrillig commented on client/my-sites/marketplace/components/billing-interval-switcher/index.tsx:
```diff
@@ -48,6 +65,19 @@ const BillingIntervalSwitcher: React.FunctionComponent< Props > = ( {
 						{ annualLabel }
+						{ plugin && (
+							<PluginAnnualSaving plugin={ plugin }>
+								{ ( annualSaving: { saving: never } ) =>
```
The type of this argument is confusing to me. `never` means that it will never be set (I think), but then you're using it below. Was this supposed to be `string | null`, maybe?
```suggestion
								{ ( annualSaving: { saving: string | null } ) =>
```

lupus2k replied:
You are right, `string | null` is the correct definition

sirbrillig commented on client/my-sites/marketplace/components/billing-interval-switcher/index.tsx:
```diff
@@ -57,7 +87,23 @@ const BillingIntervalSwitcher: React.FunctionComponent< Props > = ( {
+			<PlansIntervalToggle intervalType={ billingPeriod } onChange={ onChange }>
+				{ plugin && (
+					<PluginAnnualSaving plugin={ plugin }>
+						{ ( annualSaving: { saving: never } ) =>
```
Same question about `never` here.

lupus2k replied:
Fixed [here](https://github.com/Automattic/wp-calypso/pull/61341/commits/95d991c37417fa7afe7935b02472a95270739605)

lupus2k committed:
Update client/my-sites/marketplace/components/billing-interval-switcher/index.tsx

- update plugin object type

Co-authored-by: Payton Swick <payton@foolord.com>

lupus2k committed:
Update client/my-sites/marketplace/components/billing-interval-switcher/index.tsx

Co-authored-by: Payton Swick <payton@foolord.com>

lupus2k committed:
Update client/my-sites/marketplace/components/billing-interval-switcher/index.tsx

Co-authored-by: Payton Swick <payton@foolord.com>

sirbrillig commented on client/my-sites/marketplace/components/billing-interval-switcher/index.tsx:
```diff
@@ -57,7 +87,23 @@ const BillingIntervalSwitcher: React.FunctionComponent< Props > = ( {
+			<PlansIntervalToggle intervalType={ billingPeriod } onChange={ onChange }>
+				{ plugin && (
+					<PluginAnnualSaving plugin={ plugin }>
+						{ ( annualSaving: { saving: never } ) =>
```
```suggestion
						{ ( annualSaving: { saving: string | null } ) =>
```

lupus2k committed:
Marketplace: Advertise the discount on annual plans
- Update ESLint import/order
- Update prop to match ESLint prettier
- Change `never` to `string | null` for `saving`

allilevine reviewed (approved):
The discount is shown as expected, and matches the design. ✅ 

We might want to revisit when we show this or the discount amounts, I'm not sure $1.00 is much of an incentive 🙂 cc @cpapazoglou 
<img width="361" alt="Screen Shot 2022-02-24 at 11 35 51 AM" src="https://user-images.githubusercontent.com/1689238/155568103-129d7a7c-6aa0-4578-b175-5f20eb6b6966.png">

cpapazoglou reviewed (dismissed):
Let's avoid showing the discount when there aren't actual savings

<img width="1145" alt="CleanShot 2022-02-24 at 19 00 24@2x" src="https://user-images.githubusercontent.com/12430020/155571641-e08d0c17-fb31-406b-b1e7-53d2e7d1563e.png">

cpapazoglou commented:
> The discount is shown as expected, and matches the design. ✅
> 
> We might want to revisit when we show this or the discount amounts, I'm not sure $1.00 is much of an incentive 🙂 cc @cpapazoglou <img alt="Screen Shot 2022-02-24 at 11 35 51 AM" width="361" src="https://user-images.githubusercontent.com/1689238/155568103-129d7a7c-6aa0-4578-b175-5f20eb6b6966.png">

Curious to know which is that product. I have found that Shipment Tracking, has the same price for monthly and annually for EURO but seems ok for USD ($4 monthly / $44 annually). 

@emmnyc85 do we want to revise the EURO pricing of Shipment Tracking?

allilevine commented:
> Curious to know which is that product.

@cpapazoglou It's WooCommerce Xero: plugins/woocommerce-xero/

lupus2k committed:
Marketplace: Advertise the discount on annual plans
- Show saving when > 0

lupus2k commented:
I [updated it](https://github.com/Automattic/wp-calypso/pull/61341/commits/236db87db4b68ed73c2bab7ea4dfa67b5c3e5311) to show the savings when the savings is more than 0.
In the case of a discount of $1.00, we have to be cautious about other currencies because $1.00 in other currencies can be a big saving.

allilevine commented on client/my-sites/plugins/plugin-saving/index.jsx:
```diff
@@ -0,0 +1,24 @@
+			productMonthly && productYearly
+				? Math.round( productMonthly.cost * 12 - productYearly.cost )
+				: null;
+		return totalDiscount > 0 ? formatCurrency( totalDiscount, productYearly.currency_code ) : null;
```
Can we use && here?
```suggestion
		return totalDiscount > 0 && formatCurrency( totalDiscount, productYearly.currency_code );
```

lupus2k committed:
Change ? to && in total Discount

Co-authored-by: Allison Levine <1689238+allilevine@users.noreply.github.com>

lsl reviewed (approved):
LGTM!

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7162223

Thank you @lupus2k for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7162223).
